### PR TITLE
[dev.operator-integrations] Reconcile operator integrations

### PR DESCRIPTION
**Pull request by rfratto**
Related to #1414 

Implements the reconcile loop and produces a working end-to-end example.

This is a massive PR; a lot of work has been done around refactoring and sharing as much code as possible to reduce bloating up the operator codebase even further. 

I have done a somewhat-decent job and breaking up the changes across commits to make them more easily reviewable; you may find reviewing the PR commit-by-commit easier than reviewing the final diffs.

<details>
<summary>Example deployment of CRDs</summary>

```yaml
apiVersion: monitoring.grafana.com/v1alpha1
kind: GrafanaAgent
metadata:
  name: grafana-agent-integrations-example
  namespace: default
  labels:
    app: grafana-agent-integrations-example
spec:
  image: grafana/agent:main
  logLevel: info
  serviceAccountName: grafana-agent
  storage:
    volumeClaimTemplate:
      spec:
        resources:
          requests:
            storage: 1Gi
  logs:
    instanceSelector:
      matchLabels:
        agent: grafana-agent-integrations-example
  metrics:
    instanceSelector:
      matchLabels:
        agent: grafana-agent-integrations-example
  integrations:
    selector:
      matchLabels:
        agent: grafana-agent-integrations-example

---

apiVersion: monitoring.grafana.com/v1alpha1
kind: MetricsInstance
metadata:
  name: primary
  namespace: default
  labels:
    agent: grafana-agent-integrations-example
spec:
  remoteWrite:
  - url: http://prometheus:9090/api/v1/write
  podMonitorNamespaceSelector: {}
  podMonitorSelector:
    matchLabels:
      instance: primary

---

apiVersion: monitoring.grafana.com/v1alpha1
kind: LogsInstance
metadata:
  name: primary
  namespace: default
  labels:
    agent: grafana-agent-integrations-example
spec:
  clients:
  - url: http://loki:8080/loki/api/v1/push

  # Supply an empty namespace selector to look in all namespaces.
  podLogsNamespaceSelector: {}
  podLogsSelector:
    matchLabels:
      instance: primary

---

# Have the Agent monitor itself.
apiVersion: monitoring.coreos.com/v1
kind: PodMonitor
metadata:
  name: grafana-agents
  namespace: default
  labels:
    instance: primary
spec:
  selector:
    matchLabels:
      app.kubernetes.io/name: grafana-agent
  podMetricsEndpoints:
    - port: http-metrics

---

# Have the Agent get logs from itself.
apiVersion: monitoring.grafana.com/v1alpha1
kind: PodLogs
metadata:
  name: grafana-agents
  namespace: default
  labels:
    instance: primary
spec:
  selector:
    matchLabels:
      app.kubernetes.io/name: grafana-agent
  pipelineStages:
    - cri: {}

---

# Collect node_exporter metrics.
apiVersion: monitoring.grafana.com/v1alpha1
kind: Integration
metadata:
  name: node-exporter
  namespace: default
  labels:
    agent: grafana-agent-integrations-example
spec:
  name: node_exporter
  type:
    allNodes: true
    unique: true
  config:
    autoscrape:
      enable: true
      metrics_instance: default/primary
    rootfs_path: /default/node_exporter/rootfs
    sysfs_path: /default/node_exporter/sys
    procfs_path: /default/node_exporter/proc
  volumeMounts:
    - mountPath: /default/node_exporter/proc
      name: proc
    - mountPath: /default/node_exporter/sys
      name: sys
    - mountPath: /default/node_exporter/rootfs
      name: root
  volumes:
    - name: proc
      hostPath:
        path: /proc
    - name: sys
      hostPath:
        path: /sys
    - name: root
      hostPath:
        path: /root
  # These aren't really used here, but just being mounted for local testing.
  secrets:
  - name: fake-secret
    key: key
  configMaps:
  - name: fake-configmap
    key: foo

---

# Collect kubernetes API events.
apiVersion: monitoring.grafana.com/v1alpha1
kind: Integration
metadata:
  name: eventhandler
  namespace: default
  labels:
    agent: grafana-agent-integrations-example
spec:
  name: eventhandler
  type:
    unique: true
  config:
    logs_instance: default/primary
    cache_path: "/var/lib/grafana-agent/data/eventhandler.cache"

---

#
# Pretend Secrets/ConfigMaps
#

---
apiVersion: v1
kind: Secret
metadata:
  name: fake-secret
  namespace: default
stringData:
  key: "value"

---
apiVersion: v1
kind: ConfigMap
metadata:
  name: fake-configmap
  namespace: default
data:
  foo: "bar"

#
# Extra resources
#

---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: grafana-agent
  namespace: default
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: grafana-agent
rules:
- apiGroups:
  - ""
  resources:
  - nodes
  - nodes/proxy
  - nodes/metrics
  - services
  - endpoints
  - pods
  - events # needed for eventhandler integration
  verbs:
  - get
  - list
  - watch
- nonResourceURLs:
  - /metrics
  - /metrics/cadvisor
  verbs:
  - get
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: grafana-agent
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: grafana-agent
subjects:
- kind: ServiceAccount
  name: grafana-agent
  namespace: default
```

</details>

### CODE REVIEW

File: pkg/operator/build_hierarchy.go [hunk 1]
```diff
@@ -248,11 +248,18 @@ func buildSecrets(ctx context.Context, cli client.Client, deploy gragent.Deploym
...
+					binaryValue, binaryFound = o.BinaryData[ref.Reference.ConfigMap.Key]
+				)
+
+				if dataFound {
+					value = dataValue
+				} else if binaryFound {
+					value = string(binaryValue)
+				} else {
 					return fmt.Errorf("no key %s in ConfigMap %s", ref.Reference.ConfigMap.Key, o.Name)
 				}
```
Comment (rfratto) [inline pkg/operator/build_hierarchy.go:262]: Needs test

File: pkg/operator/operator.go [hunk 2]
```diff
@@ -185,12 +185,14 @@ func New(l log.Logger, c *Config) (*Operator, error) {
 		For(&gragent.GrafanaAgent{}, builder.WithPredicates(agentPredicates...)).
 		Owns(&apps_v1.StatefulSet{}).
 		Owns(&apps_v1.DaemonSet{}).
+		Owns(&apps_v1.Deployment{}).
```
Comment (rfratto) [inline pkg/operator/operator.go:188]: Needs documentation update for RBAC

File: pkg/operator/operator.go [hunk 3]
```diff
@@ -185,12 +185,14 @@ func New(l log.Logger, c *Config) (*Operator, error) {
...
 		Owns(&core_v1.Secret{}).
 		Owns(&core_v1.Service{}).
 		Watches(&source.Kind{Type: &core_v1.Secret{}}, notifierHandler).
 		Watches(&source.Kind{Type: &gragent.LogsInstance{}}, notifierHandler).
 		Watches(&source.Kind{Type: &gragent.PodLogs{}}, notifierHandler).
 		Watches(&source.Kind{Type: &gragent.MetricsInstance{}}, notifierHandler).
+		Watches(&source.Kind{Type: &gragent.Integration{}}, notifierHandler).
```
Comment (rfratto) [inline pkg/operator/operator.go:195]: Needs documentation update for RBAC

File: pkg/operator/resources_integrations.go [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+		Template: tmpl,
+	}
+
+	return &apps_v1.Deployment{
+		ObjectMeta: metadataFromPodTemplate(name, d, tmpl),
+		Spec:       spec,
+	}, nil
+}
+
+func integrationsPodTemplateOptions(name string, d gragent.Deployment) podTemplateOptions {
```
Comment (rfratto) [inline pkg/operator/resources_integrations.go:55]: Needs an e2e test

File: pkg/operator/build_hierarchy.go (hunk 1 above)
Comment (rfratto) [reply pkg/operator/build_hierarchy.go:262]: Covered by the e2e test now

File: pkg/operator/apis/monitoring/v1alpha1/types_integrations.go [hunk 5]
```diff
@@ -71,39 +71,51 @@ type IntegrationSpec struct {
...
+	// as the Integration; use the secrets and configMaps fields instead.
 	Volumes []corev1.Volume `json:"volumes,omitempty"`
 
 	// An extra list of VolumeMounts to be associated with the Grafana Agent pods
 	// running this integration. VolumeMount names will be mutated to be unique
-	// for an agent deployment.
+	// across all used IntegrationSpecs.
+	//
+	// Mount paths should include the namespace/name of the Integration CR to
+	// avoid potentially colliding with order resources.
```
Comment (rlankfo) [inline pkg/operator/apis/monitoring/v1alpha1/types_integrations.go:87]: what are order resources?

Comment (rfratto) [reply pkg/operator/apis/monitoring/v1alpha1/types_integrations.go:87]: haha, oops, should say "other"

File: pkg/operator/reconciler_integrations.go [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+	deploy, err := newIntegrationsDeployment(r.config, name, d)
+	if err != nil {
+		return fmt.Errorf("failed to generate integrations Deployment: %w", err)
+	}
+	key := types.NamespacedName{Namespace: deploy.Namespace, Name: deploy.Name}
+
+	if len(d.Integrations) == 0 {
+		level.Info(l).Log("msg", "deleting integrations Deployment", "deploy", key)
+		var deploy apps_v1.Deployment
+		return deleteManagedResource(ctx, r.Client, key, &deploy)
```
Comment (rlankfo) [inline pkg/operator/reconciler_integrations.go:78]: Can you clarify why the calls to deleteManagedResource are always necessary when this condition is true? Is it possible that `len(d.Integrations)` was never > 0 when this block of code is called?

**Review by rlankfo — APPROVED** [review]
LGTM :+1:

**Pull request merged**